### PR TITLE
Rdm 2224 As a user I want to resume my draft case

### PR DESCRIPTION
**Pull request by andrewfolga**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDM-2224

### Change description ###
As a user I want to resume my draft case

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/ccd/domain/service/startevent/DefaultStartEventOperation.java [hunk 1]
```diff
@@ -64,41 +76,70 @@ public StartEventTrigger triggerStartForCaseType(final String uid,
...
 
-        final CaseDetails caseDetails = getCaseDetails(jurisdictionId, caseTypeId, caseReference);
+        final CaseDetails caseDetails = caseDetailsSupplier.get();
 
-        validateEventTrigger(() -> !eventTriggerService.isPreStateValid(caseDetails.getState(), eventTrigger));
+        validateEventTrigger(() -> !eventTriggerService.isPreStateEmpty(eventTrigger));
 
-        final String eventToken = eventTokenService.generateToken(uid, caseDetails, eventTrigger,
-            caseType.getJurisdiction(), caseType);
+        // TODO: we may need to take care of drafts that are saved for existing case so token needs to include the relevant draft payload
```
Comment (vlaurin) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/startevent/DefaultStartEventOperation.java]: That's an interesting point... I'm not sure how the drafts would fit in the existing event token flow.

File: src/test/java/uk/gov/hmcts/ccd/domain/service/common/TestBuildersUtil.java [hunk 2]
```diff
@@ -375,7 +380,7 @@ private CaseTypeBuilder() {
             caseType.setJurisdiction(new Jurisdiction());
         }
 
-        public static CaseTypeBuilder aCaseType() {
+        public static CaseTypeBuilder anCaseType() {
```
Comment (vlaurin) [inline src/test/java/uk/gov/hmcts/ccd/domain/service/common/TestBuildersUtil.java]: What's the motivation behind that change? the name seemed better before.

File: src/main/java/uk/gov/hmcts/ccd/domain/model/draft/CaseDraft.java [hunk 3]
```diff
@@ -53,4 +58,19 @@ public void setCaseDataContent(CaseDataContent caseDataContent) {
...
 
+    @JsonIgnore
+    public SecurityClassification getSecurityClassification() {
+        String securityClassification = caseDataContent != null ? caseDataContent.getSecurityClassification() : null;
+        return securityClassification != null ? SecurityClassification.valueOf(securityClassification) : null;
+    }
+
+    @JsonIgnore
+    public Map<String, JsonNode> getData() {
+        return caseDataContent != null ? caseDataContent.getData() : null;
```
[the remark this answers is not in the document]
Comment (andrewfolga) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/draft/CaseDraft.java]: Done.

File: src/main/java/uk/gov/hmcts/ccd/domain/model/draft/CaseDraft.java [hunk 4]
```diff
@@ -53,4 +58,19 @@ public void setCaseDataContent(CaseDataContent caseDataContent) {
...
+    }
+
+    @JsonIgnore
+    public Map<String, JsonNode> getDataClassification() {
+        return caseDataContent != null ? caseDataContent.getDataClassification() : null;
```
[the remark this answers is not in the document]
Comment (andrewfolga) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/draft/CaseDraft.java]: Done.

File: src/main/java/uk/gov/hmcts/ccd/domain/model/draft/CaseDraft.java [hunk 5]
```diff
@@ -53,4 +58,19 @@ public void setCaseDataContent(CaseDataContent caseDataContent) {
         this.caseDataContent = caseDataContent;
     }
 
+    @JsonIgnore
+    public SecurityClassification getSecurityClassification() {
+        String securityClassification = caseDataContent != null ? caseDataContent.getSecurityClassification() : null;
+        return securityClassification != null ? SecurityClassification.valueOf(securityClassification) : null;
```
[the remark this answers is not in the document]
Comment (andrewfolga) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/draft/CaseDraft.java]: Done.

File: src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/DefaultGetCaseViewFromDraftOperation.java [hunk 6]
```diff
@@ -62,7 +64,34 @@ public CaseView execute(String jurisdictionId, String caseTypeId, String draftId
...
+            CaseViewEvent createEvent = new CaseViewEvent();
+            createEvent.setEventId("Draft created");
+            createEvent.setEventName("Draft created");
+            createEvent.setStateId("Draft");
+            createEvent.setStateName("Draft");
+            createEvent.setUserId("");
+            createEvent.setTimestamp(draftResponse.getCreated());
+            events.add(createEvent);
+        }
+        return events.toArray(new CaseViewEvent[0]);
```
[the remark this answers is not in the document]
Comment (andrewfolga) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/DefaultGetCaseViewFromDraftOperation.java]: Bad merge I suppose. Done now.

File: src/main/java/uk/gov/hmcts/ccd/domain/service/startevent/ClassifiedStartEventOperation.java [hunk 7]
```diff
@@ -30,16 +47,46 @@ public StartEventTrigger triggerStartForCaseType(String uid, String jurisdiction
...
+                                                                                          caseTypeId));
+    }
+
+    private StartEventTrigger deduceDefaultClassificationsForDraft(StartEventTrigger startEventTrigger, String caseTypeId) {
+        CaseDetails caseDetails = startEventTrigger.getCaseDetails();
+        deduceDefaultClassificaitonIfCaseDetailPresent(caseTypeId, caseDetails);
+        return startEventTrigger;
+    }
+
+    private void deduceDefaultClassificaitonIfCaseDetailPresent(String caseTypeId, CaseDetails caseDetails) {
```
[the remark this answers is not in the document]
Comment (andrewfolga) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/startevent/ClassifiedStartEventOperation.java]: :) done

File: src/test/java/uk/gov/hmcts/ccd/domain/service/common/TestBuildersUtil.java (hunk 2 above)
Comment (andrewfolga) [reply src/test/java/uk/gov/hmcts/ccd/domain/service/common/TestBuildersUtil.java]: Codacy complains there cannot be a single letter word at the start of method name.

File: src/main/java/uk/gov/hmcts/ccd/domain/service/startevent/DefaultStartEventOperation.java (hunk 1 above)
Comment (andrewfolga) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/startevent/DefaultStartEventOperation.java]: Let's discuss once we actually need that. The comment was just to make us aware.

File: src/main/java/uk/gov/hmcts/ccd/domain/model/draft/CaseDraft.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (vlaurin) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/draft/CaseDraft.java]: Classifications are optional for callback responses; if they are omitted then we need to generate them ourselves from the defaults. So by the time we get to the `CaseDraft`, we should always have classifications no?

File: src/test/java/uk/gov/hmcts/ccd/domain/service/common/TestBuildersUtil.java (hunk 2 above)
Comment (vlaurin) [reply src/test/java/uk/gov/hmcts/ccd/domain/service/common/TestBuildersUtil.java]: Well, that's a pain :D Should we switch to `newCaseType()` or `caseType()` then instead, to try and keep it grammatically correct at least?

Comment (andrewfolga) [reply src/test/java/uk/gov/hmcts/ccd/domain/service/common/TestBuildersUtil.java]: Sure done

File: src/main/java/uk/gov/hmcts/ccd/domain/model/draft/CaseDraft.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (andrewfolga) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/draft/CaseDraft.java]: Yes true. Will amend.

File: src/main/java/uk/gov/hmcts/ccd/data/draft/DefaultDraftGateway.java [hunk 8]
```diff
@@ -154,6 +154,7 @@ private String getUriWithQueryParams() {
     }
 
     private DraftResponse assembleDraft(Draft getDraft) {
+        //TODO whenever a data structure changes we get stuck here, better ignore corrupt one and move next
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/data/draft/DefaultDraftGateway.java]: do we have to implement this TODO as part of this PR?

**Pull request merged**